### DISCUSSION TITLE
Dynamic object creation and granting other roles access to those objects / automated `snowddl apply` invocations

### DISCUSSION THREAD

**Discussion by acaruso7** [littleK0i/SnowDDL#96] [Q&A]
## Dynamic Object Creation

I have a need to create new schemas on the fly, with names that aren't known beforehand

I know I can use a tech role to do this, but then how do I grant access to other roles also managed by SnowDDL?

For example, suppose I have the following tech role:

```yaml
mydb_schema_creator:
  grants:
    DATABASE:CREATE SCHEMA:
      - mydb
```

Suppose I grant this tech role to a pre-existing business role:

```yaml
transformer:
  tech_roles:
    - mydb_schema_creator
```

Obviously if `TRANSFORMER__B_ROLE` creates some schemas and objects within those schemas, it will also have read access on them because it will be the owner. However, suppose I also want to grant access to some other business role? Since the schemas aren't managed by SnowDDL, I can't simply add `schema_read` on other business roles

## Automated `snowddl apply` invocations

There was some similar discussion in https://github.com/littleK0i/SnowDDL/discussions/58, only asking about database creation privileges (which in this case I don't think is possible with a tech role). The suggestion was to use Python config to dynamically create the dbs. This is possible of course but I don't really want to be invoking `snowddl apply` in an automated context. For example, suppose I have a CI workflow where some objects with names determined by the context of that workflow must be generated on the fly. I can read those names in with python config and then invoke `snowddl apply`, but this does not seem like best practice because of the potential side effects that may be experienced from the apply. Who knows what DDL is going to be executed? This problem is mitigated by the fact that we can apply only safe operations, and scope down to certain object types, but there are still limitations (see https://github.com/littleK0i/SnowDDL/discussions/90)

Is there any guidance anyone can share here? Is the general recommendation to _only_ use SnowDDL in a context where the plans are reviewed by a human, or can automated invocations a feasible and safe option?

**Comment by littleK0i** [comment]
Creating new schema involves quite a lot of SQL commands. You may see all of them by adding a new schema and running `snowddl plan`.

1. Create managed schema.
2. Create schema roles.
3. Add future grants.
4. Possibly grant newly created schema roles to business roles with matched patterns in `schema_owner`, `schema_write`, `schema_read`.

In theory, you may copy-paste these commands, wrap into Python function and run every time you need to create a new schema. Commands should be executed by SnowDDL admin role for consistency.

With this approach you'll emulate what SnowDDL is doind without running SnowDDL every time.

---

Running SnowDDL in "safe" mode each time should be fine as well. All commands creating new schema are considered "safe".

**Comment by littleK0i** [comment]
@acaruso7 , maybe it is a bit late, but let's give it a try.

Please note, since SnowDDL version 0.27 it is now possible to define permission model which allows creation of custom schemas.

Relevant guides:
https://docs.snowddl.com/guides/permission-model
https://docs.snowddl.com/basic/yaml-configs/permission-model

Specific example for Fivetran: https://docs.snowddl.com/guides/other-guides/fivetran

Even if we internally never allow external software to create schemas, I recognize this as a perfectly valid use case. So this feature is now fully supported.

**Comment by julesmga** [comment]
@littleK0i I have a similar use case, and the DATABASE_OWNER ruleset covers most of my needs, however I'd like something a little different and I'm not sure how to go about it.

For more context, we are using SQLMesh to which I gave the database owner role so it can create schemas. SQLMesh has an interesting feature which creates virtual environments for our developers, meaning they too need to be able to create schemas on the fly. For now they've been given database owner as well, but this makes for a really cluttered snowflake as everybody can see all virtual environments (that's a lot of schemas). I would like for our devs to be able to create schemas, but the schemas would ideally belong to the role who created them (as well as the database owner role so sqlmesh can still see them as it can clean up old schemas when they are no longer needed).

Any idea how I could achieve this with snowddl? I tried creating a technical role who can create schemas, but the devs can't use them at all since they belong to the database owner.

**Reply by littleK0i** [reply]
Yeah, this specific setup is a bit difficult in Snowflake, since `CREATE SCHEMA` is not enough. You also need to run a huge amount of `GRANT` and `GRANT ON FUTURE` commands to make it work.

But there is an alternative. As far as I know, SQLMesh can change "catalog" (database) name depending on environment. Also, instead of creating schemas it can use object names suffixes. In theory, it should be possible to have this:

```
PROD_DB.PROD_SCHEMA   -- one prod env, clean object names

DEV_DB.DEV1_SCHEMA    -- all envs of specific developer, objects with suffixes
DEV_DB.DEV2_SCHEMA
DEV_DB.DEV3_SCHEMA
```

In this case SnowDDL can create all schemas and correct permissions beforehand. While individual developers will be able to see only their own DEV environments and nothing else.

Also, it will keep number of DEV schemas at reasonable level. Even if individuals devs may not see see schemas of other devs, admin can see everything. With unrestricted schema creation it may become problematic quite fast.

**Reply by julesmga** [reply]
Sounds exactly like what I want, will have a look, thx!

**Reply by julesmga** [reply]
Works like a charm, thanks for the tip. We now have a sparkling clean production database!
